external/llvm-project/mlir/lib/Conversion/AsyncToGPU/AsyncToGPU.cpp

```diff
@@ -0,0 +1,363 @@
+      b.setInsertionPoint(oprAllocOp);
+    /// hack: add one global wait for gpu.allocs
+    auto allocWait = makeWait(b, loc);
+    auto gpuMemType = opr.getType();
+    auto dst = b.create<gpu::AllocOp>(loc, gpuMemType, tokenType,
+                                      ValueRange{allocWait}, ValueRange{},
+                                      ValueRange{});
+    auto dstMem = dst.getResult(0);
+    auto dstToken = dst.getResult(1);
+    // if alloc, convert to gpu.alloc
+    if (oprAllocOp) {
+      // TODO(sjw): make sure accessors are all on the GPU
+      oprAllocOp->replaceAllUsesWith(ValueRange{dstMem});
+    } else {
+      if (func::isAccessModeRead(accessMode)) {
+        // else copy to device
+        auto asyncDeps = ValueRange{dstToken};
+        auto memcpyToken =
+            b.create<gpu::MemcpyOp>(loc, tokenType, asyncDeps, dstMem, opr);
+        dstToken = memcpyToken.getResult(0);
+      }
+      if (func::isAccessModeWrite(accessMode)) {
+        copyBackOprs[fidx] = dstMem;
+      }
+    }
+    asyncDeps.push_back(dstToken);
+    return dstMem;
+  }
+
+  LogicalResult
+  matchAndRewrite(async::LaunchOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+    auto module = op->getParentOfType<ModuleOp>();
+    auto ctx = module.getContext();
+
+    assert(op->getNumResults() == 1); // only 1 async.token
+
+    // 1. get target{gpu} attribute from func
+
+    auto gpuAttr = getGPUTarget(op);
+    if (!gpuAttr.hasValue())
+      return op.emitOpError("requires a gpu target");
+
+    auto arch = gpuAttr->get("arch");
+    auto binary = gpuAttr->get("binary");
+    auto blockSize = gpuAttr->get("block_size").template cast<IntegerAttr>();
+    auto gridSize = gpuAttr->get("grid_size").template cast<IntegerAttr>();
+
+    auto func = *getCalledFunc(op);
+    auto floc = func.getLoc();
+
+    // Also capture the accessMap for the params, default all to read-write
+    SmallVector<func::AccessMode, 8> accessVec(func.getNumArguments(),
+                                               func::AccessMode::ReadWrite);
+    if (auto accessMap =
+            func->template getAttrOfType<ArrayAttr>("access_map")) {
+      std::transform(
+          accessMap.begin(), accessMap.end(), accessVec.begin(),
+          [](Attribute a) {
+            return a.template cast<func::AccessModeAttr>().getValue();
+          });
+    }
+
+    // 2. create dummy gpu.module
+    //    - with gpu.binary, arch attributes
+    //    - and llvm.func (may not be necessary
+    //    gpu.module @<func_name>_module attributes {arch = "gfx908", gpu.binary
+    //    = "\7FELF\..."} {
+    //      llvm.func @<func_name> (...) attributes {block_size = 256 : i32,
+    //      gpu.kernel, grid_size = 900 : i32, rocdl.kernel}
+    //        // does any of that matter?
+
+    FunctionOpInterface funcIF(func);
+    auto funcName = funcIF.getName();
+    auto gpuModuleName = funcName + "_module";
```

Comment:
Check for collisions, add a suffix if needed.

Reply:
It does check for collision, but re-uses the same GPU module/func when found.